src/collapse/recursive.rs

```diff
@@ -0,0 +1,215 @@
+
+    fn set_nthreads(&mut self, n: usize) {
+        self.opt.nthreads = n;
+    }
+}
+
+impl Folder {
+    fn line_parts(line: &str) -> Option<(&str, usize)> {
+        let mut parts = line.rsplitn(2, ' ');
+        let count = parts.next()?.parse().ok()?;
+        let stack = parts.next()?;
+        Some((stack, count))
+    }
+
+    fn collapse_stack(stack: Cow<str>) -> Cow<str> {
+        // First, determine whether we can avoid allocation by just returning
+        // the original stack (in the case that there is no recursion, which is
+        // likely the mainline case).
+        if !Self::is_recursive(&stack) {
+            return stack;
+        }
+
+        // There is recursion, so we can't get away without allocating a new
+        // String.
+        let mut result = String::with_capacity(stack.len());
+        let mut last = None;
+        for frame in stack.split(';') {
+            match last {
+                None => {
+                    result.push_str(frame);
+                    result.push(';')
+                }
+                Some(l) => {
+                    if l != frame {
+                        result.push_str(frame);
+                        result.push(';')
+                    }
+                }
+            }
+            last = Some(frame);
+        }
+
+        // Remove the trailing semicolon
+        result.pop();
+
+        result.into()
+    }
+
+    /// Determine whether or not a stack contains direct recursion.
+    fn is_recursive(stack: &str) -> bool {
+        let mut last = None;
+        for current in stack.split(';') {
+            match last {
+                None => {
+                    last = Some(current);
+                }
+                Some(l) => {
+                    if l == current {
+                        // Recursion!
+                        return true;
+                    } else {
+                        last = Some(current);
+                    }
+                }
+            }
+        }
+        false
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use super::*;
+
+    #[test]
+    fn test_collapse_stack() {
+        assert_eq!(Folder::collapse_stack("".into()), "");
+        assert_eq!(Folder::collapse_stack("single".into()), "single");
+        assert_eq!(
+            Folder::collapse_stack("not;recursive".into()),
+            "not;recursive"
+        );
+        assert_eq!(
+            Folder::collapse_stack("has;some;some;recursion;recursion".into()),
+            "has;some;recursion"
+        );
+        assert_eq!(
+            Folder::collapse_stack("co;recursive;co;recursive".into()),
+            "co;recursive;co;recursive"
+        );
+    }
+
+    #[test]
+    fn test_line_parts() {
+        assert_eq!(
+            Folder::line_parts("foo;bar;baz 42"),
+            Some(("foo;bar;baz", 42))
+        );
+        assert_eq!(
+            Folder::line_parts("something; including spaces  42"),
+            Some(("something; including spaces ", 42))
```

Comment:
The trailing space here is questionable — I genuinely don't know whether the collapsed format dictates that the count is only ever offset by a _single_ space. Maybe avoid testing that specifically here, and just keep a single space before 42?